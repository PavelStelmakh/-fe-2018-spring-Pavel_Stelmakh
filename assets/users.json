[
  {
    "id": 1,
    "name": "Pasha",
    "age": "19",
    "password": "1111",
    "dateOfBirth": "2011-10-05T00:00:00.000Z",
    "dateOfFirstLogin": "2011-10-05T14:48:00.000Z",
    "dateOfNextNotif": "2011-10-05T14:48:00.000Z",
    "information": "meaw"
  },
  {
    "id": 2,
    "name": "Sasha",
    "age": "24",
    "password": "2222",
    "dateOfBirth": "2011-10-05T00:00:00.000Z",
    "dateOfFirstLogin": "2011-10-05T14:48:00.000Z",
    "dateOfNextNotif": "2011-10-05T14:48:00.000Z",
    "information": "meaw"
  },
  {
    "id": 3,
    "name": "Masha",
    "age": "20",
    "password": "3333",
    "dateOfBirth": "2011-10-05T00:00:00.000Z",
    "dateOfFirstLogin": "2011-10-05T14:48:00.000Z",
    "dateOfNextNotif": "2011-10-05T14:48:00.000Z",
    "information": "meaw"
  }
]