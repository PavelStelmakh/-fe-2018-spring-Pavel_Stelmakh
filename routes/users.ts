import { Router, Response, Request, NextFunction } from 'express';
const router = Router();
import { User } from '../models/User';
const loginRoute = require('./login');
const users = require('../models/Users');
import { getUserByToken } from '../shared/getUserByToken';

router.use('/login', loginRoute);

router.use('*', (req: Request, res: Response, next: NextFunction) => {
    const data = getUserByToken(req);

    if (data.user) {
        next();
    } else {
        res.status(401);
        res.end();
    }
});

router.get('/users', (_req: Request, res: Response) => {
    res.status(200).send(users.users());
});

router.get('/users/search/:name', (req: Request, res: Response) => {
    try {
        const name: number = req.params['name'];
        const searchUsers: User[] = users.search(name);
        setTimeout(() => {
            if (searchUsers.length > 0) {
                res.status(200).send(searchUsers);
            } else {
                res.status(400);
                res.end();
            }
        }, 1000);
    }
    catch (Error) {
        res.status(418);
        res.end();
    }
});

router.get('/users/:id', (req: Request, res: Response) => {
    try {
        const id: number = +req.params['id'];
        const user: User | undefined = users.user(id);
        if (user) {
            setTimeout(() => res.status(200).send(user), 3000);
        } else {
            res.status(400);
            res.end();
        }
    }
    catch (Error) {
        res.status(418);
        res.end();
    }
});

router.post('/users/add', (req: Request, res: Response) => {
    try {
        const user: User = users.add(req.body);
        res.status(200).send(user);
    }
    catch (Error) {
        res.status(418);
    }
    res.end();
});

router.put('/users/:id', (req: Request, res: Response) => {
    try {
        const id: number = +req.params['id'];
        const success: boolean = users.update(id, req.body);
        if (success) {
            res.status(200);
        } else {
            res.status(418);
        }
    }
    catch (Error) {
        res.status(500);
    }
    res.end();
});

router.delete('/users/:id', (req: Request, res: Response) => {
    try {
        const id: number = +req.params['id'];
        const success: boolean = users.delete(id);
        if (success) {
            res.status(200);
        } else {
            res.status(418);
        }
    }
    catch (Error) {
        res.status(500);
    }
    res.end();
});

module.exports = router;