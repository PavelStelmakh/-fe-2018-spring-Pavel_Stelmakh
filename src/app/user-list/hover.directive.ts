import { Directive, ElementRef, Renderer2 } from '@angular/core';

@Directive({
  selector: '[appHover]',
  host: {
    '(mouseenter)': 'onMouseEnter()',
    '(mouseleave)': 'onMouseLeave()'
  }
})
export class HoverDirective {

  constructor(private element: ElementRef, private renderer: Renderer2) {          
    this.renderer.setStyle(this.element.nativeElement, 'cursor', 'pointer');
  }

  onMouseEnter() {
    this.setHover('rgba(102, 102, 102, 0.5)');
  }

  onMouseLeave() {
    this.setHover('rgba(102, 102, 102, 0)');
  }

  private setHover(val: string) {
    this.renderer.setStyle(this.element.nativeElement, 'background-color', val);
  }

}
