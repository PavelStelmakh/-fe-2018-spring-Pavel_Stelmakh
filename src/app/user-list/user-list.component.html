<app-user-chosen [user]="user" 
(expand)="expandList($event)" 
(delete)="updateUsers()"
(edit)="editUser($event)"></app-user-chosen>
<div *ngIf="expand">
    <input type="text" placeholder="search by name" [formControl]="username" (focus)="focusInput()">
    <div (click)="searchUsers()" [class.iconSearch]="true"><fa-icon [icon]="faSearch" size="1x" style="color: #ccc"></fa-icon></div>
    <app-user-dropdown-list *ngIf="focus">
        <app-user-dropdown-item *ngFor="let user of users" [user]="user" (click)="selectUser(user)" appHover></app-user-dropdown-item>
        <div *ngIf="notFound">{{'user-list.notFound' | translate}}</div>
    </app-user-dropdown-list>
</div>