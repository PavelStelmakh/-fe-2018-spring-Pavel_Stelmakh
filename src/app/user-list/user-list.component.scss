@import '../shared/driveAnimation';

@include animation("list", 0, 20vh, rgba(0, 0, 0, 0), rgba(0, 0, 0, 0.7));

:host {    
    text-align: center;
    position: relative;
    display: block;
    width: 500px;
    color: #ccc;
}

input[type="text"] {
    width: 300px;
    margin-bottom: 10px
}

.iconSearch {
    display: inline-block;
    height: 27px;
}