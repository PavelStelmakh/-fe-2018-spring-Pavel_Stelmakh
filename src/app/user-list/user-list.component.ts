import { Component, OnInit, Output, EventEmitter } from '@angular/core';
import { Store, select } from '@ngrx/store';
import { User } from '../../../models/User';
import { faSearch, IconDefinition } from '@fortawesome/free-solid-svg-icons';
import { FormControl } from '@angular/forms';
import { skip } from 'rxjs/operators';
import * as usersAction from '../actions/users.action';
import * as usersReducer from '../reducers/users.reducer';

@Component({
  selector: 'app-user-list',
  templateUrl: './user-list.component.html',
  styleUrls: ['./user-list.component.scss']
})
export class UserListComponent implements OnInit {
  @Output() edit: EventEmitter<void> = new EventEmitter<void>();
  username: FormControl;
  faSearch: IconDefinition = faSearch;
  users: User[];
  expand: boolean;
  focus: boolean;
  notFound: boolean;
  updateListUsers: boolean;

  constructor(private store: Store<usersReducer.State>) {}

  ngOnInit() {
    this.users = [];
    this.updateListUsers = true;
    this.username = new FormControl('');
    this.focus = false;
    this.store.pipe(select(usersReducer.getUsersArray), skip(1)).subscribe(
      (users: User[]) => {
        if (this.users.length !== users.length) {
          this.updateListUsers = true;
        }
        this.users = users;
        if (users.length === 0) {
          this.notFound = true;
        } 
      }
    );
  }

  selectUser(user: User) {
    this.store.dispatch(new usersAction.SelectUserAction(user.id));
    this.focus = false;
  }

  focusInput() {
    this.notFound = false;
    this.focus = true;
    if(this.updateListUsers) {
      this.users = [];
      this.store.dispatch(new usersAction.LoadUsersAction('all'));
      this.updateListUsers = false;
    }
  }

  expandList(expand: boolean) {
    this.focus = false;
    this.expand = expand;
    this.username.setValue('');
  }

  searchUsers() {
    this.store.dispatch(new usersAction.LoadUsersAction(this.username.value));
  }

  updateUsers() {
    this.username.setValue('');
    this.users = [];
  }

  editUser() {
    this.edit.emit();
  }

}
