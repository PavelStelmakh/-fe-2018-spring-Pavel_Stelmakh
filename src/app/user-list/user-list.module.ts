import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FontAwesomeModule } from '@fortawesome/angular-fontawesome';
import { ReactiveFormsModule } from '@angular/forms';
import { UserListComponent } from './user-list.component';
import { UserChosenComponent } from './user-chosen/user-chosen.component';
import { HttpClientModule, HttpClient } from '@angular/common/http';
import { TranslateModule, TranslateLoader } from '@ngx-translate/core';
import { HttpLoaderFactory } from '../shared/HttpLoaderFactory';
import { HoverDirective } from './hover.directive';
import { UserDropdownListComponent } from './user-dropdown-list/user-dropdown-list.component';
import { UserDropdownItemComponent } from './user-dropdown-item/user-dropdown-item.component';

@NgModule({
  imports: [
    CommonModule,
    ReactiveFormsModule,
    FontAwesomeModule,
    HttpClientModule,
    TranslateModule.forRoot({
      loader: {
        provide: TranslateLoader,
        useFactory: HttpLoaderFactory,
        deps: [HttpClient]
      }
    })
  ],
  declarations: [
    UserListComponent, 
    UserChosenComponent,
    HoverDirective,
    UserDropdownListComponent,
    UserDropdownItemComponent
  ],
  providers: [
  ],
  exports: [UserListComponent]
})
export class UserListModule { }
