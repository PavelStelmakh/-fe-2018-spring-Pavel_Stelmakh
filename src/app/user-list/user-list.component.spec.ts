import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { NO_ERRORS_SCHEMA } from '@angular/core';
import { Observable, of } from 'rxjs';
import { TranslateModule, TranslateLoader, TranslateService } from '@ngx-translate/core';
import { ReactiveFormsModule } from '@angular/forms';
import { Store } from '@ngrx/store';
import { hot } from 'jasmine-marbles';
import { UserListComponent } from './user-list.component';
import * as usersReducer from '../reducers/users.reducer';

const translations = {
  "user-list": {
    "notFound": "sorry, nothing is found"
  } 
};

const users = [];

class FakeLoader implements TranslateLoader {
  getTranslation(lang: string): Observable<{}> {
    return of(translations);
  }
}

describe('UserListComponent', () => {
  let fixture: ComponentFixture<UserListComponent>;
  let component: UserListComponent;
  let translate: TranslateService;
  let store: Store<usersReducer.State>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [
        TranslateModule.forRoot({
          loader: {provide: TranslateLoader, useClass: FakeLoader}
        }),
        ReactiveFormsModule
      ],
      declarations: [ 
        UserListComponent
       ],
      providers: [ 
        {
          provide: Store,
          useValue: {
            dispatch: jest.fn(),
            pipe: jest.fn(() => hot('-a', { a: users }))
          }
        }
       ],
       schemas: [
        NO_ERRORS_SCHEMA
       ]
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(UserListComponent);
    translate = TestBed.get(TranslateService);
    component = fixture.componentInstance;
    store = TestBed.get(Store);
    translate.use('en');
  });
  
  it('renders markup to snapshot with empty array of users', async () => {
    component.expand = true;
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

});
