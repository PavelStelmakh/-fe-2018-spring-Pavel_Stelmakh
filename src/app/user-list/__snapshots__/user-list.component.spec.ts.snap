// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserListComponent renders markup to snapshot with empty array of users 1`] = `
<app-user-list
  edit={[Function EventEmitter]}
  expand={[Function Boolean]}
  faSearch={[Function Object]}
  focus="false"
  store={[Function Object]}
  updateListUsers={[Function Boolean]}
  username={[Function FormControl]}
  users={[Function Array]}
>
  <app-user-chosen /><div>
    <input
      class="ng-untouched ng-pristine ng-valid"
      ng-reflect-form="[object Object]"
      placeholder="search by name"
      type="text"
    />
    <div
      class="iconSearch"
    >
      <fa-icon
        size="1x"
        style="color: #ccc"
      />
    </div>
    
  </div>
</app-user-list>
`;
